### PR TITLE
Batch r nuget: commands

### PR DESCRIPTION
**Pull request by KevinRansom**

### CODE REVIEW

File: Microsoft.DotNet.Interactive.CSharp/CSharpKernel.cs [hunk 1]
```diff
@@ -352,14 +352,14 @@ bool ShouldRebuild()
         }
 
         public async Task LoadExtensionsFromDirectory(
-            DirectoryInfo directory,
+            string directory,
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.CSharp/CSharpKernel.cs]: Why the change from `DirectoryInfo`?

File: Microsoft.DotNet.Interactive.CSharp/CSharpKernelExtensions.cs [hunk 2]
```diff
@@ -57,115 +80,146 @@ public static class CSharpKernelExtensions
...
+            command.Handler = CommandHandler.Create<PackageReference, KernelInvocationContext>(async (package, pipelineContext) =>
+            {
+                var addPackage = new AddPackage(package)
                 {
-                    var message = $"Installing package {package.PackageName}";
-                    if (!string.IsNullOrWhiteSpace(package.PackageVersion))
+                    Handler = async context =>
                     {
-                        message += $", version {package.PackageVersion}";
+                        await Task.FromResult(
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.CSharp/CSharpKernelExtensions.cs]: There's no need for this to be `async`.

File: Microsoft.DotNet.Interactive.Tests/LanguageKernelTests.cs [hunk 3]
```diff
@@ -959,40 +960,33 @@ public async Task When_SubmitCode_command_adds_packages_to_kernel_then_PackageAd
...
-
-            events
-                .Should()
-                .Contain(e => e is DisplayedValueUpdated && ((DisplayedValueUpdated)e).Value.ToString().Contains("done!"));
+                .ContainSingle(e => e is DisplayedValueProduced && ((DisplayedValueProduced)e).Value.ToString().Contains("Installing package"));
 
             events
                 .Should()
-                .ContainSingle(e => e is NuGetPackageAdded);
+                .ContainSingle(e => e is DisplayedValueUpdated && ((DisplayedValueUpdated)e).Value.ToString().Contains("done!"));
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.Tests/LanguageKernelTests.cs]: This looks like a much weaker assertion than checking for the specific event type.

File: Microsoft.DotNet.Interactive/Commands/LoadExtensionsInDirectory.cs [hunk 4]
```diff
@@ -11,15 +11,15 @@ namespace Microsoft.DotNet.Interactive.Commands
     public class LoadExtensionsInDirectory : KernelCommandBase
     {
         public LoadExtensionsInDirectory(
-            DirectoryInfo directory, 
-            IReadOnlyList<FileInfo> additionalDependencies = null)
+            string directoryPath,
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/Commands/LoadExtensionsInDirectory.cs]: I'd prefer not to weaken the type safety here. What's the reasoning?

File: Microsoft.DotNet.Interactive/Commands/NugetRestoreDirective.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+namespace Microsoft.DotNet.Interactive.Commands
+{
+    public class NugetRestoreDirective : KernelCommandBase
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/Commands/NugetRestoreDirective.cs]: Command naming is typically imperative, e.g. `RestoreNugetPackage`.

File: Microsoft.DotNet.Interactive/KernelExtensionLoader.cs [hunk 6]
```diff
@@ -17,7 +17,7 @@ public class KernelExtensionLoader
             FileInfo assemblyFile, 
             IKernel kernel, 
             KernelInvocationContext context,
-            IEnumerable<FileInfo> additionalDependencies = null)
+            IEnumerable<string> additionalDependencies = null)
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/KernelExtensionLoader.cs]: I think `additionalDependencies` should be able to be removed in favor of something easier to understand. Not sure if you've seen an opportunity to do that but don't be worried about breaking existing extension functionality because it needs to be redone anyway.

File: Microsoft.DotNet.Interactive/PackageRestoreContext.cs [hunk 7]
```diff
@@ -35,78 +49,98 @@ public PackageRestoreContext()
...
-            string restoreSources = null)
+            string restoreSources = null,
+            string sourceCode = null)
         {
-            var requestedPackage = new NugetPackageReference(packageName, packageVersion, restoreSources);
+            var requestedPackage = new PackageReference(packageName, packageVersion, restoreSources, sourceCode);
+            var packageKey = requestedPackage.PackageKey;
 
-            if (!string.IsNullOrEmpty(packageName) && _nugetPackageReferences.TryGetValue(requestedPackage, out _))
+            lock (_packageReferences)
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/PackageRestoreContext.cs]: Rather than needing a `lock`, you can use `ConcurrentDictionary`.

File: Microsoft.DotNet.Interactive.CSharp/CSharpKernel.cs (hunk 1 above)
Comment (KevinRansom) [reply Microsoft.DotNet.Interactive.CSharp/CSharpKernel.cs]: Because NewtonSoft json doesn't serialize, FileInfo, DirectoryInfo.

File: Microsoft.DotNet.Interactive.Tests/LanguageKernelTests.cs (hunk 3 above)
Comment (KevinRansom) [reply Microsoft.DotNet.Interactive.Tests/LanguageKernelTests.cs]: Immediately below is this assertion, which accurately verifies the event:
This line verifies that the updated output text is produced.

````
            events.OfType<PackageAdded>()
                  .Should()
                  .BeEquivalentTo(new NugetPackageReference("Microsoft.Extensions.Logging", "2.2.0"));
                  .ContainSingle(e => ((PackageAdded)e).PackageReference.PackageName == "Microsoft.Extensions.Logging" 
                                   && ((PackageAdded)e).PackageReference.PackageVersion == "2.2.0");

````

File: Microsoft.DotNet.Interactive/Commands/LoadExtensionsInDirectory.cs (hunk 4 above)
Comment (KevinRansom) [reply Microsoft.DotNet.Interactive/Commands/LoadExtensionsInDirectory.cs]: DirectoryInfo and FileInfo do not serialize by newtonsoft json.

File: Microsoft.DotNet.Interactive/Commands/NugetRestoreDirective.cs (hunk 5 above)
Comment (KevinRansom) [reply Microsoft.DotNet.Interactive/Commands/NugetRestoreDirective.cs]: K.

File: Microsoft.DotNet.Interactive/PackageRestoreContext.cs (hunk 7 above)
Comment (KevinRansom) [reply Microsoft.DotNet.Interactive/PackageRestoreContext.cs]: Agreed

File: Microsoft.DotNet.Interactive.CSharp/CSharpKernelExtensions.cs (hunk 2 above)
Comment (KevinRansom) [reply Microsoft.DotNet.Interactive.CSharp/CSharpKernelExtensions.cs]: Unfortunately Handler is defined on KernelBase as:
````
    public delegate Task KernelCommandInvocation(
        KernelInvocationContext context);
````
Therefore in order to avoid the warning:  "async method saying that it lacks 'await' operators"
we need away to add an await.

File: Microsoft.DotNet.Interactive/PackageRestoreContext.cs [hunk 8]
```diff
@@ -141,14 +152,18 @@ public Task<ResolvedNugetPackageReference> GetResolvedNugetPackageReference(stri
...
-                       .Select(xs => new ResolvedNugetPackageReference(
+                       .Select(xs => new ResolvedPackageReference(
                                    xs.Key.packageName,
                                    xs.Key.packageVersion,
                                    xs.Select(x => x.assemblyPath).ToArray(),
                                    xs.Key.packageRoot,
                                    probingPaths))
                        .ToDictionary(r => r.PackageName, StringComparer.OrdinalIgnoreCase);
 
+            lock (_lockObject)
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/PackageRestoreContext.cs]: > lock (_lockObject) [](start = 11, length = 19)

This lock looks unnecessary.

File: Microsoft.DotNet.Interactive/Utility/PackageRestoreResult.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.DotNet.Interactive
+{
+    public class PackageRestoreResult
+    {
+        public PackageRestoreResult(
+            bool succeeded,
+            IEnumerable<PackageReference> requestedPackages,
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/Utility/PackageRestoreResult.cs]: >IEnumerable<PackageReference> requestedPackages, [](start = 12, length = 48)

This is unused.

File: Microsoft.DotNet.Interactive.Tests/LanguageKernelTests.cs [hunk 10]
```diff
@@ -1275,6 +1269,292 @@ public async Task Dependency_version_conflicts_are_resolved_correctly()
             events.Should().NotContainErrors();
         }
 
+        [Fact]
+        public async Task PoundRNugetDisallowsEmptyPackageSpecification()
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.Tests/LanguageKernelTests.cs]: >PoundRNugetDisallowsEmptyPackageSpecification [](start = 26, length = 45)

Test naming quibble: snake case please.

File: Microsoft.DotNet.Interactive/Commands/RestoreNugetDirective.cs [hunk 11]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+namespace Microsoft.DotNet.Interactive.Commands
+{
+    public class RestoreNugetDirective : KernelCommandBase
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/Commands/RestoreNugetDirective.cs]: >RestoreNugetDirective [](start = 17, length = 21)

Would RestorePackages capture the intent here? You've removed "Nuget" from other class names, and "Directive" sounds like the direct object here, which isn't accurate.

**Comment by KevinRansom** [comment]
@jonsequitur 
It tells me I can't merge the PR.  Nor can I resolve the conflicts.

I will resolve them On my PC and push another commit.

image

File: Microsoft.DotNet.Interactive/Events/PackageAdded.cs [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    public class PackageAdded : KernelEventBase
+    {
+        public PackageAdded(AddPackage addPackage): base(addPackage)
+        {
+            if (addPackage == null)
+            {
+                throw new System.ArgumentNullException(nameof(addPackage));
+            }
+
+            PackageReference = addPackage.PackageReference;
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive/Events/PackageAdded.cs]: It's preferable to take in the `PackageReference` directly here. Then long as the parameter name matches the property name (case-insensitive), the deserialization will work correctly.

File: Microsoft.DotNet.Interactive.Tests/LanguageKernelTests.cs [hunk 13]
```diff
@@ -1287,7 +1567,12 @@ public async Task issue_637()
 //using System.Text.Json;
 ");
 
-            events.Should().NotContainErrors();
+            events
+                .OfType<StandardErrorValueProduced>()
+                .Last()
+                .Value
+                .Should()
+                .Be("");
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.Tests/LanguageKernelTests.cs]: Why this change? This doesn't look like the correct behavior.

Comment (KevinRansom) [reply Microsoft.DotNet.Interactive.Tests/LanguageKernelTests.cs]: Hmm, it isn't ...

**Comment by KevinRansom** [comment]
@jonsequitur 
GitHub hates me:

image
